cpp/src/spatial/hausdorff.cu

```diff
@@ -26,89 +26,78 @@
-        int space_b_end   =                        space_offsets[space_b_idx];
+        return;
+    }
 
-        T min_dist_sqrd = 1e20;
+    int space_a_idx   = block_idx % num_spaces;
+    int space_a_begin = space_a_idx == 0 ? 0 : space_offsets[space_a_idx - 1];
+    int space_a_end   =                        space_offsets[space_a_idx];
 
-        int num_points_in_b = space_b_end - space_b_begin;
+    int space_b_idx   = block_idx / num_spaces;
+    int space_b_begin = space_b_idx == 0 ? 0 : space_offsets[space_b_idx - 1];
+    int space_b_end   =                        space_offsets[space_b_idx];
 
-        if (threadIdx.x < num_points_in_b)
-        {
-            T point_b_x = xs[space_b_begin + threadIdx.x];
-            T point_b_y = ys[space_b_begin + threadIdx.x];
+    T min_dist_sqrd = 1e20;
 
-            for (int i = space_a_begin; i < space_a_end; i++)
-            {
-                T point_a_x = xs[i];
-                T point_a_y = ys[i];
-                T dist_sqrd = magnitude_squared(point_b_x - point_a_x, point_b_y - point_a_y);
+    int num_points_in_b = space_b_end - space_b_begin;
 
-                min_dist_sqrd = min(min_dist_sqrd, dist_sqrd);
-            }
-        }
+    if (threadIdx.x < num_points_in_b)
+    {
+        T point_b_x = xs[space_b_begin + threadIdx.x];
+        T point_b_y = ys[space_b_begin + threadIdx.x];
 
-        if (min_dist_sqrd > 1e10)
+        for (int i = space_a_begin; i < space_a_end; i++)
         {
-            min_dist_sqrd = -1;
-        }
-
-        __shared__ T dist_sqrd[1024];
+            T point_a_x = xs[i];
+            T point_a_y = ys[i];
+            T dist_sqrd = magnitude_squared(point_b_x - point_a_x, point_b_y - point_a_y);
 
-        dist_sqrd[threadIdx.x] = -1;
-
-        __syncthreads();
-
-        if (threadIdx.x < num_points_in_b)
-        {
-            dist_sqrd[threadIdx.x] = min_dist_sqrd;
+            min_dist_sqrd = min(min_dist_sqrd, dist_sqrd);
         }
+    }
 
-        __syncthreads();
+    if (min_dist_sqrd > 1e10)
+    {
+        min_dist_sqrd = -1;
+    }
 
-        for (int offset = blockDim.x / 2; offset > 0; offset >>= 1)
-        {
-            if (threadIdx.x < offset)
-            {
-                dist_sqrd[threadIdx.x] = max(dist_sqrd[threadIdx.x],
-                                             dist_sqrd[threadIdx.x + offset]);
-            }
+    __shared__ T dist_sqrd[1024];
 
-            __syncthreads();
-        }
+    dist_sqrd[threadIdx.x] = threadIdx.x < num_points_in_b ? min_dist_sqrd : -1;
 
+    for (int offset = blockDim.x / 2; offset > 0; offset >>= 1)
+    {
         __syncthreads();
 
-        if (threadIdx.x == 0)
+        if (threadIdx.x < offset)
         {
-            results[block_idx] = (dist_sqrd[0] < 0) ? 1e10 : sqrt(dist_sqrd[0]);
+            dist_sqrd[threadIdx.x] = max(dist_sqrd[threadIdx.x],
+                                            dist_sqrd[threadIdx.x + offset]);
         }
     }
+
```

Comment:
You need a __syncthreads() after the loop before you read the shared memory from a different thread, since you moved it to the top.

Reply:
Read is on the same thread as the write: 0.

Reply:
Read is on the same thread as the write, no?

Reply:
This is because in the final iteration of the loop `offset` will be 1, so only thread 0 will write to `dist_sqrd[0]`, followed by only thread 0 reading from `dist_sqrd[0]`, right?

Reply:
That's my understanding.

Reply:
I guess this is all going to be replaced by a CUB call in the future anyway.